File: JavaScript Application/src/simulate.js
const ds = require('./datastore');

function wait(ms) { return new Promise(res => setTimeout(res, ms)); }

async function worker(name, ops) {
  for (const op of ops) {
    if (op.type === 'add') {
      const t = await ds.addTask({ title: `${op.title} (by ${name})`, category: op.category, assignedTo: op.assignedTo, createdBy: name });
      console.log(`[${name}] added`, t.id);
    } else if (op.type === 'complete') {
      await wait(op.delay || 0);
      const t = await ds.completeTask(op.id);
      console.log(`[${name}] completed`, op.id, '->', !!t);
    } else if (op.type === 'assign') {
      const t = await ds.assignTask(op.id, op.user);
      console.log(`[${name}] assigned`, op.id, 'to', op.user);
    }
  }
}

async function main() {
  console.log('Simulating concurrent users...');
  const w1 = worker('Alice', [
    { type: 'add', title: 'Buy milk', category: 'Home', assignedTo: 'Alice' },
    { type: 'add', title: 'Prepare slides', category: 'Work', assignedTo: 'Bob' },
  ]);

  const w2 = worker('Bob', [
    { type: 'add', title: 'Fix bug', category: 'Work', assignedTo: 'Bob' },
    { type: 'complete', id: 1, delay: 300 },
  ]);

  await Promise.all([w1, w2]);
  console.log('Simulation complete. Current tasks:');
  const tasks = await ds.listTasks({});
  tasks.forEach(t => console.log(`${t.id}. [${t.completed ? 'x' : ' '}] ${t.title} - ${t.assignedTo || '—'}`));
}

main().catch(err => { console.error(err); process.exit(1); });

File: Java Application/src/com/todo/Simulate.java
package com.todo;

import com.todo.core.Datastore;
import com.todo.models.Task;
import java.util.List;

public class Simulate {

    static class Worker extends Thread {
        private Runnable[] ops;

        public Worker(String threadName, Runnable... ops) {
            super(threadName);
            this.ops = ops;
        }

        @Override
        public void run() {
            try {
                for (Runnable op : ops) {
                    op.run();
                    Thread.sleep(100);
                }
            } catch (InterruptedException e) { }
        }
    }

    public static void start() throws Exception {
        System.out.println("Simulating concurrent users...");

        Worker w1 = new Worker("Alice", 
            () -> {
                Task t = Datastore.addTask("Buy milk (by Alice)", "Home", "Alice", "Alice");
                System.out.println("[Alice] added " + t.getId());
            },
            () -> {
                Task t = Datastore.addTask("Prepare slides (by Bob)", "Work", "Bob", "Alice");
                System.out.println("[Alice] added " + t.getId());
            }
        );

        Worker w2 = new Worker("Bob", 
            () -> {
                Task t = Datastore.addTask("Fix bug (by Bob)", "Work", "Bob", "Bob");
                System.out.println("[Bob] added " + t.getId());
            },
            () -> {
                try { Thread.sleep(300); } catch (InterruptedException e) {}
                Task t = Datastore.completeTask(1);
                System.out.println("[Bob] completed 1 -> " + (t != null));
            }
        );

        w1.start();
        w2.start();
        
        w1.join();
        w2.join();
        
        System.out.println("Simulation complete. Current tasks:");
        List<Task> tasks = Datastore.listTasks(null, null, null);
        for (Task t : tasks) {
            String check = t.isCompleted() ? "x" : " ";
            String assigned = t.getAssignedTo() != null ? t.getAssignedTo().getName() : "-";
            System.out.printf("%d. [%s] %s - %s\n", t.getId(), check, t.getTitle(), assigned);
        }
    }
}
